File: contrib/hexspeak.py
#!/usr/bin/env python3

import re
import sys


def solve_recursive_count(words, currentLen, used, targetLength, cnt):
    for i in range(1, targetLength - currentLen + 1):
        for word in words[i]:
            if word in used:
                continue
            if i != targetLength - currentLen:
                solve_recursive_count(
                    words, currentLen + i, used + [word], targetLength, cnt)
            else:
                cnt[0] += 1


def solve_nonrecursive_count(words, targetLength):
    from collections import deque
    candidates = deque([([], 0)])
    cnt = 0
    while candidates:
        wordsSoFar, currentLen = candidates.popleft()
        if currentLen == targetLength:
            cnt += 1
        else:
            for i in range(1, targetLength - currentLen + 1):
                for word in words.get(i, []):
                    if word not in wordsSoFar:
                        candidates.append(
                            (wordsSoFar + [word], currentLen + i))
    return cnt


def get_words_per_length(dictionaryFile, letters):
    words = [[] for _ in range(128)]
    m = re.compile(r'^[' + letters + ']*$')
    for word in open(dictionaryFile):
        word = word.strip()
        if len(word) > 2 and m.match(word):
            if word in ['aaa', 'aba', 'abc']:
                continue
            if word not in words[len(word)]:
                words[len(word)].append(word)
    words[1] = ['a']
    return words


def main():
    if len(sys.argv) != 4:
        targetLength = 8
        letters = 'abcdef01'
        dictionaryFile = '/usr/share/dict/words'
    else:
        targetLength = int(sys.argv[1])
        letters = sys.argv[2].replace('0', 'o').replace('1', 'il')
        dictionaryFile = sys.argv[3]
    words = get_words_per_length(dictionaryFile, letters)
    cnt = [0]
    import time
    start = time.time()
    solve_recursive_count(words, 0, [], targetLength, cnt)
    end = time.time()
    print(cnt[0], "in", 1000*(end-start), "ms")
    #
    # Not using recursion is much slower! Apparently cache is thrashed
    #
    # results = solve_nonrecursive_count(words, targetLength)
    # print "Total:", results
    # import resource # Proof this has much more memory impact...
    # print resource.getrusage(resource.RUSAGE_SELF).ru_maxrss, "KB"

if __name__ == "__main__":
    main()

File: contrib/Nim/hexspeak.nim
#!/usr/bin/env python3

import os
import re
import times
import strutils
import sets

type WordCounts = array[128, seq[ptr string]]
type UsedWords = HashSet[ptr string]

# Probably violating Nim here: to optimize the speed of execution,
# like in C++, I "intern" the strings. Basically, when loading the
# words, I store them in this global container - and in the UsedWords
# container (mutated during the recursion below) I maintain the
# *pointers* to the words - instead of the words themselves. This
# means that when checking whether the container has seen a word before
# or not, the comparison compares pointers instead of strings.
#
# Much faster!
#
# But if a GC happens, we're hosed - ergo, why I believe I am violating
# all Nim's tenets by doing this :-) Am I? (see the uses of "addr"
# below to take the addresses of strings in this sequence.
var allWords : seq[string]

proc get_words_per_length(dictionaryFile:string, letters:string): WordCounts =
  var regexStr = r"^[" & letters & r"]*$"
  var m = re(regexStr)
  allwords.add("a")
  var p_a = addr allwords[0]
  for line in open(dictionaryFile).lines():
    var word = line.strip()
    if len(word) > 2 and match(word, m):
      if word in ["aaa", "aba", "abc"]:
        continue
      allWords.add(word)
      result[len(word)].add(addr allWords[^1])
  result[1].add(p_a)


proc solve_recursive_count(
    words: WordCounts, currentLen:int, used:ref UsedWords, targetLength:int, cnt:int):int =
  var cnt = cnt
  # echo "[-] currentLen:", currentLen, ", cnt:", cnt, ", used:", used.join(",")
  for i in countup(1, targetLength - currentLen):
    for word in words[i]:
      if word in used[]:
        continue
      if i != targetLength - currentLen:
        used[].incl(word)
        cnt = solve_recursive_count(
          words, currentLen + i, used, targetLength, cnt)
        used[].excl(word)
      else:
        inc cnt
  cnt


proc main =
  var targetLength: int
  var letters: string
  var dictionaryFile: string
  if paramCount() != 3:
    targetLength = 8
    letters = "abcdef01"
    dictionaryFile = "/usr/share/dict/words"
  else:
    targetLength = paramStr(1).parseInt
    letters = paramStr(2).replace("0", "o").replace("1", "il")
    dictionaryFile = paramStr(3)
  var words = get_words_per_length(dictionaryFile, letters)
  var used = new UsedWords
  var start = now()
  var cnt = solve_recursive_count(words, 0, used, targetLength, 0)
  var stop = now()
  echo cnt, " in ", (stop-start).inMilliseconds, " ms."

main()
